api/include/opentelemetry/context/context.h

```diff
@@ -0,0 +1,205 @@
+  // of the context object linked list.
+  Context(nostd::string_view key, ContextValue value)  
+  {
+    head_ = BuildHead(key, value);
+    next_ = NULL;
+  }
+
+  // Accepts a new iterable and then returns a new context that
+  // contains the new key and value data.
+  template <class T , nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
+  Context* SetValues(T &values) noexcept
+  {
+    Context *new_node = new Context(values, this);
+    return new_node;
+  }
+  
+  // Accepts a new iterable and then returns a new context that
+  // contains the new key and value data.
+  Context* SetValue(nostd::string_view key, ContextValue value) noexcept
+  {
+    Context *new_node = new Context(key,value, this);
+    return new_node;
+  }
+
+  // Returns the value associated with the passed in key.
+  common::AttributeValue GetValue(nostd::string_view key)
+  {
+    Context *context_trace = this;
+    
+    //Iterate through the context nodes
+    while(context_trace != NULL){
+      DataList *data_trace = context_trace->head_;
+      
+      //Iterate through the internal data nodes
+      while(data_trace != NULL){
+        if(strncmp(key.data(), data_trace->key_, 50) == 0){
+          return data_trace->value_;
+        }
+        
+        data_trace = data_trace->next_;
+      }
+      
+      context_trace = context_trace->next_;
+    }
+    
+    return "";
+  }
+
+  // Iterates through the current and comparing context objects
+  // to check for equality.
+  bool operator==(const Context &other)
+  {
+    // If the two contexts are not at the place in the list
+    // they cannot be equivalent
+    if(next_ != other.next_){
+      return false; 
+    }
+    
+    DataList *trace = head_;
+    DataList *other_trace = other.head_;
+    
+    // Iterate through the lists and compare the keys
+    // TODO should also compare the values but the nostd::variant
+    // seems to not have a comparison operator
+    while(trace != NULL){
+
+      if(other_trace == NULL){
+        return false;
+      }
+      
+      if((trace->key_ != other_trace->key_)){
+        return false;
+      }
+
+      trace = trace->next_;
+      other_trace = other_trace->next_;
+    }
+
+    return true;
+  }
+
+  Context(const Context &other) = default;
+  
+  Context &operator=(const Context &other){
+    head_ = other.head_;
+    next_ = other.next_;
+  }
+
+private:
+  
+  // Head of the list which holds the keys and values of this context 
+  DataList *head_;
+  
+  // Pointer to the next context object in the context list
+  Context *next_;
+  
+  // Builds a data list off of a key and value iterable and returns the head 
+  template <class T , nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
+  DataList* BuildList(const T &keys_and_vals){
+    auto iter = std::begin(keys_and_vals);
+    
+    // Create list head 
+    DataList *head = new DataList;
```

Comment:
When are DataList de-allocated?

Keep in mind that Context's copy constructor and assignment operator reuse head across multiple contexts. Let me know if you don't see good options.

Reply:
It is deallocated within the context destructor and the copy and assignment constructors now copy the other contexts data list into their own. It seems like by allowing copying and assignment though, that it is becoming less like a list and a little more like a tree, but I don't think that it is an issue.